src/company.cpp

```diff
@@ -0,0 +1,157 @@
+#include "company.h"
+#include <vector>
+#include <algorithm>
+#include <iostream>
+#include <set>
+#include <limits.h>
+#include <cassert>
+
+Worker::Worker() {}
+
+Worker::Worker(unsigned long start, unsigned long finish, unsigned long type, unsigned long salary, unsigned long id) {
+	w_start = start;
+	w_finish = finish;
+	assert(w_start < w_finish);
+	w_type = type;
+	w_salary = salary;
+	w_id = id;
+}
+
+unsigned long Worker::get_type() {
+	return w_type;
+}
+
+unsigned long Worker::get_start() {
+	return w_start;
+}
+
+unsigned long Worker::get_finish() {
+	return w_finish;
+}
+
+unsigned long Worker::get_salary() {
+	return w_salary;
+}
+
+bool operator<(const Worker w_1, const Worker w_2) {
+	return w_1.w_start <= w_2.w_start;
+}
+
+std::istream& operator>> (std::istream &is, Worker &w) {
+	is >> w.w_start;
+	is >> w.w_finish;
+	is >> w.w_type;
+	is >> w.w_salary;
+	is >> w.w_id;
+	return is;
+}
+
+Task::Task() {}
+
+Task::Task(unsigned long type, unsigned long cost) {
+	t_type = type;
+	t_cost = cost;
+}
+
+bool operator<(const Task t_1, const Task t_2) {
+	return t_1.t_type <= t_2.t_type;
+}
+
+bool Task::operator==(const Task& other) const {
+	return t_type == other.t_type && t_cost == other.t_cost;
+}
+
+unsigned long Task::get_cost() {
+	return t_cost;
+}
+
+unsigned long Task::get_type() {
+	return t_type;
+}
+
+std::istream &operator>> (std::istream &is, Task &t) {
+	is >> t.t_type;
+	is >> t.t_cost;
+	return is;
+}
+
+std::ostream &operator<< (std::ostream &os, const Task &t) {
+	os << t.t_type;
+	return os;
+}
+
+Manager::Manager() {}
+
+Manager::Manager(unsigned long start, unsigned long id) {
+	m_start = start;
+	m_id = id;
+}
+
+unsigned long Manager::get_id() {
+	return m_id;
+}
+
+unsigned long Manager::get_start() {
+	return m_start;
+}
+
+bool operator<(const Manager m_1, const Manager m_2) {
+	return m_1.m_start <= m_2.m_start;
+}
+
+
+std::istream &operator>> (std::istream &is, Manager &m) {
+	is >> m.m_start;
+	is >> m.m_id;
+	return is;
+}
+
+Result::Result() {
+	r_value = ULONG_MAX;
+}
+
+void Result::calculate_result(std::vector<Task> Tasks, std::vector<Manager> Managers, std::vector<Worker> Workers) {
```

Comment:
May be we can split Result::calculate_result into several helpful methods?
1) Let's assume Result is if fact just the wrapper of mangers' tasks order: 
`std::vector<std::pair<Manager, std::vector<Task>> tasks_order`

2) we can introduce a helper method which uses std::next_permutation to loop over all possible permutations and generates stream of all possible Results (method 'next' of stream generates next Result object)

3) Then we can ask Result about be the cost for arbitrary input workers schedule.
long Result::calculate(std::vector<Worker> Workers) {
...
}
Using such method we can check for different workers's schedules what is the penalty value for the fixed manager's tasks order and do not keep it in Result object.
For instance it could be helpful if we change a little bit schedule of i.e. one worker and see how it changes penalty value when managers follow the fixed order of executed tasks

What do you think?

Reply:
So you suggest we have an additional class to store managers and tasks, which has 'next' method to generate next Result object based on a permutation?

Reply:
Let's discuss tomorrow